design/decisions/0005-entity-registration-api.md

```diff
@@ -0,0 +1,266 @@
+# Entity Deregister API
+
+* Date: __2024-10-01__
+* Status: __New__
+
+## Background
+
+The existing MQTT entity registration API `v1` which expects entities to publish their registration messages
+as retained messages to their respective entity topic ids has some limitations.
+This design expects entities not to conflict with each other in terms of their topic ids or `@id` values.
+And since there is no direct feedback to these messages, the publishing entities need to just hope that it just works.
+This expectation might be okay in a very controlled environment where the person provisioning the devices in the field
+has complete control over all the devices to avoid any conflicts and even avoid any mistakes (conflicting ids).
+But, in environments where that is not feasible, the current model without feedback can have unpleasant side effects.
+
+For example, if an entity registers itself with topic id: `te/device/child01//` and `@id`: `child01`,
+a device twin with external id `child01` is created in the cloud and all of its data is routed to this twin.
+If another device accidentally registers itself using the same topic id, but with `@id`: `child02`,
+a second device twin is created in the cloud with the external id: `child02`
+and all the messages published to that topic by `child02` and even `child01` are routed to the new device twin.
+The `child01` twin would remain orphaned in the cloud, without the user noticing anything immediately, for lack of feedback.
+
+To avoid such issues, the registering entities must be made aware of any conflicts/errors immediately
+so that they don't proceed publishing any data to the wrong entity assuming that it succeeded.
+In this case, %%te&& must have rejected the second registration message,
+unless it was explicitly done to correct a mistake in the previous registration.
```

Comment:
I would be explicit and add that "the API must distinguish requests creating an entity registration from those updating an entity registration".

Reply:
Resolved by c33517187d6c49c3e6de72f041ba53a73df0dc88